### PR TITLE
Implements ICU version of getDefaultLocale, fixes up ICU extension handling

### LINKED ISSUE #3658: (4h,P3) Intl-ICU: platform.getDefaultLocale
* [x] Implement fallback logic with default value
* [ ] Implement with ICU

### PR DESCRIPTION
**Pull request by jackhorton**
Closes #3658

### CODE REVIEW

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 1]
```diff
@@ -608,20 +637,42 @@
...
+        if (isICU) {
+            // ICU's getDefaultLocale() will return a valid BCP-47/RFC 5646 langtag
+            locale = platform.getDefaultLocale();
+            const match = platform.builtInRegexMatch(locale, /-u(-[^\-][^\-]?-[^\-]+)*(-co-[^\-]+).*/);
+            if (match) {
+                collation = callInstanceFunc(StringInstanceReplace, match[2], "-co-", "");
+                locale = callInstanceFunc(StringInstanceReplace, locale, match[2], "");
+            }
+        } else {
+            // Windows' getDefaultLocale() will return a weird RFC4646 langtag
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: lol -- "weird" is a funny way of saying something from an outdated standard, but yeah

**Comment by dilijev** [comment]
Tell me more about the fix to `supportedLocalesOf`?

File: lib/Runtime/Library/InJavascript/Intl.js (hunk 1 above)
Comment (dilijev) [reply lib/Runtime/Library/InJavascript/Intl.js]: nit: this is a quirk of WinGlob not Windows per se, but the program logic makes that clear

**Comment by jackhorton** [comment]
Sorry, it was actually an issue in resolveLocaleHelper that only surfaced in supportedLocalesOf -- it was the change to ICU appending only "-" to the string instead of "-u-" because the getExtensions JS fallback returned ["u", "co", "phonebk"] for "de-u-co-phonebk", whereas I think the WinGlob native implementation returns just ["co", "phonebk"].

**Comment by dilijev** [comment]
@jackhorton I think maybe the WinGlob platform implementation of getExtensions actually returns the "key-value" sequence. I noticed this issue as well and added a comment. Maybe handling it is now effectively fixed in your change -- we can clean up later.

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 2]
```diff
@@ -256,7 +272,20 @@
     }
 
     var strippedDefaultLocale = function () {
-        return platform.builtInRegexMatch(GetDefaultLocale(), /([^_]*).*/)[1];
+        if (isPlatformUsingICU) {
+            const def = platform.getDefaultLocale();
+            const match = platform.builtInRegexMatch(def, LANG_TAG_EXT_RE);
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >LANG_TAG_EXT_RE) [](start = 58, length = 16)

Ensure this regex is not undefined first. If it is undefined, run the initializer function function. This pattern is used elsewhere.

e.g.:
            if (!LANG_TAG_RE) {
                InitializeLangTagREs();
            }

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 3]
```diff
@@ -608,20 +667,51 @@
...
-        if (collation === undefined || availableBcpTags === undefined) { return locale; }
+        let locale = undefined;
+        let collation = undefined;
+        if (isPlatformUsingICU) {
+            // ICU's getDefaultLocale() will return a valid BCP-47/RFC 5646 langtag
+            locale = GetDefaultLocale();
+            const match = platform.builtInRegexMatch(locale, /-u(-[^\-][^\-]?-[^\-]+)*-co-([^\-]+).*/);
+            if (match) {
+                collation = match[2];
+                locale = callInstanceFunc(StringInstanceReplace, locale, `-co-${match[2]}`, "");
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >match[2] [](start = 80, length = 8)

nit (TODO): you just set collation = match[2], so use ${collation} instead

Comment (dilijev) [reply lib/Runtime/Library/InJavascript/Intl.js]: Here, do we intentionally still replace `-co-value` instead of stripping all subtags? (I'm guessing this mapping possibly involves other subtags?) 
nit (TODO): (Come to think of it `reverseLocaleAcceptingCollationValues` needs a comment or example explaining what the resulting object looks like, because I hate having to use a debugger every time I want to answer that question.)

---
In reply to: [141544228](https://github.com/Microsoft/ChakraCore/pull/3820#discussion_r141544228) [](ancestors = 141544228)

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 4]
```diff
@@ -608,20 +667,51 @@
...
+            }
+        } else {
+            // Windows' getDefaultLocale() will return a weird RFC4646 langtag
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >weird [](start = 57, length = 5)

nit (TODO): I think we should remove "weird" since we're just editorializing. It simply is an RFC4646 langtag, and weird here makes it sound like something about the tags are not quite RFC4646. Maybe maybe the distinction as: "RFC4646 (not RFC5646/BCP47) langtag"

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 5]
```diff
@@ -608,20 +667,51 @@
...
+            return __mappedDefaultLocale;
+        }
 
-        var bcpTag = availableBcpTags[collation];
-        if (bcpTag !== undefined) {
-            return locale + "-u-co-" + bcpTag;
+        const collationMapForLocale = reverseLocaleAcceptingCollationValues[locale];
+        if (collationMapForLocale === undefined) {
+            // Assume the system wouldn't give us back a bad collation value
+            __mappedDefaultLocale = `${locale}-u-co-${collation}`;
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >`${locale}-u-co-${collation}`; [](start = 36, length = 30)

Have you exercised this path with a hypothetical default locale with /-u-.*/
For some reason I think we'd be adding a duplicate singleton `-u-` which (at least according to Intl spec) is not allowed in langtags.

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 6]
```diff
@@ -608,20 +667,51 @@
...
+
+        if (collation === undefined) {
+            __mappedDefaultLocale = locale;
+            return __mappedDefaultLocale;
+        }
 
-        var bcpTag = availableBcpTags[collation];
-        if (bcpTag !== undefined) {
-            return locale + "-u-co-" + bcpTag;
+        const collationMapForLocale = reverseLocaleAcceptingCollationValues[locale];
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >reverseLocaleAcceptingCollationValues [](start = 38, length = 37)

I think the whole point of these mappings is to translate RFC 4646 (or associated standard/database/list) collation values into RFC 5646 (or associated) collation values. See the Intl explainer for the RFC 5646-associated list.
IOW this shouldn't be necessary for ICU because the langtag should already be RFC 5646.
However, since this value is in theory reported by the system, it might still be a problem. I'd like to think that ICU would convert or ignore anything it gets from Windows that isn't valid according to ICU.

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 7]
```diff
@@ -608,20 +667,51 @@
...
+        if (collationMapForLocale === undefined) {
+            // Assume the system wouldn't give us back a bad collation value
+            __mappedDefaultLocale = `${locale}-u-co-${collation}`;
+            return __mappedDefaultLocale;
         }
 
-        return locale;
+        const mappedCollation = collationMapForLocale[collation];
+        if (mappedCollation !== undefined) {
+            __mappedDefaultLocale = `${locale}-u-co-${mappedCollation}`;
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >`${locale}-u-co-${mappedCollation}` [](start = 36, length = 35)

Same -- have you exercised this? The Intl debugger should be helpful here.

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 8]
```diff
@@ -608,20 +667,51 @@
...
+        } else {
+            __mappedDefaultLocale = `${locale}-u-co-${collation}`;
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: Same -- have you exercised these paths? The Intl harness should be helpful here.

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 9]
```diff
@@ -232,10 +249,27 @@
...
+                } else {
+                    resolved += "-u-";
+                }
+            }
+
+            resolved += callInstanceFunc(ArrayInstanceJoin, subTags, "-");
+        } else {
+            resolved = undefined;
+        }
         return setPrototype({
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: nit (TODO): add a blank line before the return.

File: lib/Runtime/Base/WindowsGlobalizationAdapter.cpp [hunk 10]
```diff
@@ -829,20 +829,41 @@ if (this->object) \
...
-    int IcuIntlAdapter::GetUserDefaultLocaleName(_Out_ LPWSTR lpLocaleName, _In_ int cchLocaleName)
+    size_t IcuIntlAdapter::GetUserDefaultLanguageTag(_Out_ char16* langtag, _In_ size_t cchLangtag)
     {
-        // XPLAT-TODO (doilij): implement GetUserDefaultLocaleName
-        const auto locale = _u("en-US");
-        const size_t len = 5;
-        if (lpLocaleName)
+        UErrorCode error = UErrorCode::U_ZERO_ERROR;
+        char bcp47[ULOC_FULLNAME_CAPACITY];
+        char defaultLocaleId[ULOC_FULLNAME_CAPACITY];
```
Comment (dilijev) [inline lib/Runtime/Base/WindowsGlobalizationAdapter.cpp]: I mused about this in comments in other places -- it's not too expensive to zero-out these arrays ` = { 0 }` and it's safer in case something goes wrong.
Now that we run this function only once on Intl initialization (thanks to caching the returned value in Intl.js), it seems worth it to pay that cost here.

File: lib/Runtime/Base/WindowsGlobalizationAdapter.cpp [hunk 11]
```diff
@@ -829,20 +829,41 @@ if (this->object) \
...
+
+        int32_t written = uloc_getName(nullptr, defaultLocaleId, ULOC_FULLNAME_CAPACITY, &error);
+        if (U_FAILURE(error) || written <= 0 || written >= ULOC_FULLNAME_CAPACITY)
```
Comment (dilijev) [inline lib/Runtime/Base/WindowsGlobalizationAdapter.cpp]: > if (U_FAILURE(error) || written <= 0 || written >= ULOC_FULLNAME_CAPACITY) [](start = 7, length = 75)

nit: success branch first for readability, unless it's easier to express the failure case in an `if` conditional.
To work around this, personally I will sometimes declare a boolean named (something like) `success` or `failed` and then negate if necessary depending on what I actually want to express for the if condition.

File: lib/Runtime/Library/InJavascript/Intl.js (hunk 5 above)
Comment (jackhorton) [reply lib/Runtime/Library/InJavascript/Intl.js]: That's correct. I think this should be resolved by making ```__mappedDefaultLocale = `${locale}-co-${collation}` ```

File: lib/Runtime/Library/InJavascript/Intl.js (hunk 6 above)
Comment (jackhorton) [reply lib/Runtime/Library/InJavascript/Intl.js]: Doing a lookup in a dictionary once during the lifetime of the program is worth not needing to worry about this, I think. We may want to re-evaluate that decision if we make these dictionaries into a native-calculated, hardcoded thing (then, we could just not have this dictionary at all if ICU is enabled).

File: lib/Runtime/Library/InJavascript/Intl.js (hunk 3 above)
Comment (jackhorton) [reply lib/Runtime/Library/InJavascript/Intl.js]: As far as I know (and as far as its variable name suggests), `localesAcceptingCollationValues` is only for collation values. As for whether the other extensions need to be stripped, as of _right now_, this function is only used as the `defaultLocaleFunc` for `resolveLocales` when initializing the Collator (initializing NumberFormat and DateTimeFormat uses `strippedDefaultLocale` instead). That internally calls `platform.getExtensions`, which falls back to `getExtensionSubtags(locale)`, which is basically just 

```js
locale
  .match(LANG_TAG_EXT_RE)[0] // get extensions on the locale string
  .split('-') // split them into an array
  .filter(x => !!x) // remove empty elements
```

So, should we strip all subtags? No, probably not. Also, thinking about it, we probably shouldn't be stripping all subtags for NumberFormat or DateTimeFormat either. Id imagine the original reason for the different functions was that Collator needed the system collation value, so it went through the process of stripping "_collation" and mapping and re-formatting it, while NumberFormat and DateTimeFormat didn't need the collation, so they could blindly throw away the "_collation" part of the string, because they knew they could get -ca and -nu formats later from platform.getExtensions. In ICU, we cant do that right now and possibly don't need to do that ever. Also, since we are caching the results now, the perf cost of calculating the mapped collation value (_if we even need to_, re: the comment below) is 0 for the Number and Date case.

TL;DR: Stripping the locale for ICU is at best useless and at worst harmful, so I think we should actually ditch the ICU implementation of strippedDefaultLocale regardless

**Comment by dilijev** [comment]
@dotnet-bot test this please

**Comment by dilijev** [comment]
@mmitche is something broken?

**Comment by mmitche** [comment]
@dilijev Jenkins upgrade hit a bunch of github api rate limit issues

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 12]
```diff
@@ -590,38 +652,78 @@
         // "x-IV": ["mathan"]
     }, null);
 
-    var reverseLocaleAcceptingCollationValues = (function () {
-        var toReturn = setPrototype({}, null);
+    // reverses the keys and values in each locale's sub-object in localesAcceptingCollationValues
+    // localesAcceptingCollationValues[locale][key] = value -> reverseLocalesAcceptingCollationValues[locale][value] = key
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: Thanks for the explainer!
